app/services/sso/openid_connect.rb

```diff
@@ -25,6 +25,16 @@ def logout_url
       controller.extlogout_users_path
     end
 
+    def support_expiration?
+      true
+    end
+
+    def expiration_url
+      redirect_uri = "#{controller.main_app.extlogin_users_path}/redirect_uri"
+      "#{redirect_uri}?logout=#{CGI.escape(request.base_url + controller.main_app.extlogin_users_path)}"
```

Comment:
wouldn't this cause logout from sso? this reminds me of the same discussion we had with the `logout_url` implementation

Reply:
Yes, this would cause a logout from SSO which is correct in this case. When user logouts out from his own by clicking the Foreman `logout` button he must be redirected to the extlogout page.

But, when his access token has expired, we must use the refresh token to get new acess token from the OpenID provider. Refresh token is a feature in itself therefore, we must first implement logout from SSO and then I am working on the Refresh token part.

So for Foreman 2.0 we will provide something like this:
1) User logs in from SSO
2) User performs actions till his/her access token is valid.
3) On expiration of the access token user will be logged out of SSO. He needs to re-authenticate.

For foreman 2.1 I would propose:
1) User logs in from SSO
2) User performs actions till his/her access token is valid.
3) On expiration of the access token, a refresh token mechanism will be used to get a new pair of access token and refresh token and user will still be authenticated.

Reply:
Just to clearify, that PR was for the `logout_url` and this is for the `expiration_url`. Let me know what you think about the above comment! :)

Reply:
Foreman 2.0 has already branched over a month ago. We aren't going to pull in a new feature to it at this point.
That said, why not just redirect to `controller.main_app.extlogin_users_path` like we do in apache sso? We shouldn't be in the business of logging people out of sso - that's the sso's business.

Reply:
Clearly, I went bad with the versions there! Also, as we discussed, `mod_auth_oidc` documentation suggests that we must  perform logging out of the sso session in this manner https://github.com/zmartzone/mod_auth_openidc/wiki#9-how-do-i-logout-users. This way the `mod_auth_oidc_session` shall be removed from the cookies and new session  from the OpenID provider can be requested.

Reply:
> We shouldn't be in the business of logging people out of sso - that's the sso's business.
I was looking for some alternate method for performing this and i found this: https://github.com/zmartzone/mod_auth_openidc/wiki/Access-Tokens-and-Refresh-Tokens#access-token-revocation-and-cache-invalidation 

The doc specifies a way to eliminate the session from the cookie but this works for a very specific version ( > 2.0.1rc2) and we are using one the previous version (1.8.8). Although, I don't know if this will even work with the new version, haven't tried it out. Also, I don't know if we can get the version packaged or not? @ekohl can you maybe give some insight on what we can do?

Ideally, this is what we could do if we had a version > 2.0.1rc2 (i could be wrong, this is my understanding):

```
openid_connect.rb

def clear_cache
  url = "#{request.base_url}#{redirect_uri}?remove_at_cache=#{session[:token]}"
  uri = URI.parse(url)
  req = Net::HTTP::Get.new(uri)
  response = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true) do |https|
    https.request(req)
  end
  json_response = JSON.parse(response.body)
end

def expiration_url
  clear_cache
  controller.main_app.extlogin_users_path
end

def login_url
  clear_cache
  controller.main_app.extlogin_users_path
end
```
Or in,

```
session.rb

def session_expiry
 clear_cache if session[:sso_method] = "SSO::OpenidConnect"
 ...
 ...
end 
```
After we clear the cache, on the next call the token needs to be introspected and that action will fail because the token is not valid anymore and therefore, if will try to re-authenticate.

Also, this should do exactly the same thing that we are right not with the `?logout=` too. What do you think, is it worth the version upgrade or we can stick to what we have now?

@tbrisker let me know your thoughts.